library/std/src/net/addr.rs

```diff
@@ -290,7 +290,11 @@ impl SocketAddrV4 {
     /// ```
     #[stable(feature = "rust1", since = "1.0.0")]
     pub fn ip(&self) -> &Ipv4Addr {
-        unsafe { &*(&self.inner.sin_addr as *const c::in_addr as *const Ipv4Addr) }
+        // SAFETY:
+        //   * `in_addr` is `struct { _: [u8; 4]; }` in most OSes.
```

Comment:
"most" isn't enough. This needs to be safe on *all* OSes.

Reply:
At least [POSIX compliant systems](https://pubs.opengroup.org/onlinepubs/009695399/basedefs/netinet/in.h.html) define `in_addr` as a structure that includes at least the following member: `in_addr_t`, while `type in_addr_t = u32`.
Not sure if Rust support non-POSIX systems.

Reply:
Yes, it does -- e.g. Windows.
It would still be good to link to the POSIX docs, and say something like "non-POSIX OSes have to ensure to also satisfy this condition" or so. Maybe with a FIXME to find some more prominent place to document this.

Reply:
Yeah, Windows' `in_addr` definition: https://docs.microsoft.com/en-us/windows/win32/api/winsock2/ns-winsock2-in_addr
```c
struct in_addr {
  union {
    struct {
      u_char s_b1;
      u_char s_b2;
      u_char s_b3;
      u_char s_b4;
    } S_un_b;
    struct {
      u_short s_w1;
      u_short s_w2;
    } S_un_w;
    u_long S_addr;
  } S_un;
};
```

Anyway, I think returning `&Ipv4Addr` while the Self type doesn't contains it is a mistake.

Reply:
It's stable though.

A more explicit option would be to have a method somewhere in `std::sys` that does the conversion, so each platform has to explicitly acknowledge that this is possible. That's material for a separate PR though.

Reply:
Off-topic: What if it becomes unsafe on some platforms. Can we break stability guarantee with that reason?

Reply:
> Off-topic: What if it becomes unsound on some platforms. Can we break stability guarantee with that reason?

No, I don't think so. In the end this is a case-by-case decision, but I'd be very surprised if we did. I'd expect we would make the function panic on that platform.